### PR TITLE
fix: update monaco config even in invalid spec

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

- update monaco config (version of spec) even in invalid spec.
- adjust config of monaco to read new json schemas for spec.
- update deps like `@asyncapi/converter` and templates.

**Related issue(s)**
Resolves https://github.com/asyncapi/studio/issues/252

### CODE REVIEW

**Comment by magicmatatjahu** [comment]
@mcturco Could you look at it?

**Comment by magicmatatjahu** [comment]
@mcturco Yeah, you're right person for this, I mean for approval 😄 but if you want, you can make review of changes. I know that you probably don't understand the source code of that project but these changes are very needed.

File: package.json [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "@asyncapi/openapi-schema-parser": "^2.0.1",
     "@asyncapi/parser": "^1.15.1",
     "@asyncapi/react-component": "^1.0.0-next.38",
-    "@asyncapi/specs": "^3.0.0",
+    "@asyncapi/specs": "^2.14.0",
```
Comment (fmvilas) [inline package.json:34]: I don't think I understand why you're downgrading to a previous version of schemas. Can you elaborate on this?

File: src/services/specification.service.ts [hunk 2]
```diff
@@ -45,6 +47,12 @@ export class SpecificationService {
         return asyncApiDoc;
       })
       .catch(err => {
+        try {
+          const asyncapiSpec = YAML.load(rawSpec);
+          MonacoService.updateLanguageConfig((asyncapiSpec as { asyncapi: string }).asyncapi);
```
Comment (fmvilas) [inline src/services/specification.service.ts:52]: Can't we do something like this?

```suggestion
          const asyncapiSpec: { asyncapi: string } = YAML.load(rawSpec);
          MonacoService.updateLanguageConfig(asyncapiSpec.asyncapi);
```

Or like this if that fails:

```suggestion
          const asyncapiSpec = YAML.load(rawSpec) as { asyncapi: string };
          MonacoService.updateLanguageConfig(asyncapiSpec.asyncapi);
```

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (magicmatatjahu) [reply package.json]: Doesn't anyone read the PR description? 😄 

> switch to old json schemas due to problem that monaco cannot fetch schemas from https://asyncapi.com/definitions...

However I try to make a workaround to support latest specs.

**Comment by magicmatatjahu** [comment]
@fmvilas Could you check again?

File: package.json (hunk 1 above)
Comment (fmvilas) [reply package.json]: I did read the PR description but I still don't understand the reason.

File: package.json [hunk 3]
```diff
@@ -27,20 +27,20 @@
...
-    "@monaco-editor/react": "^4.2.2",
+    "@monaco-editor/react": "^4.4.5",
     "@semantic-release-plus/docker": "^3.1.1",
     "js-base64": "^3.7.2",
     "js-file-download": "^0.4.12",
     "js-yaml": "^4.1.0",
-    "monaco-editor": "^0.28.1",
-    "monaco-yaml": "^2.5.1",
+    "monaco-editor": "0.28.1",
+    "monaco-yaml": "2.5.1",
```
Comment (fmvilas) [inline package.json:43]: What is the reasoning behind fixing these versions?

File: package.json (hunk 1 above)
Comment (magicmatatjahu) [reply package.json]: Look on latest changes.

File: src/react-app-env.d.ts [hunk 4]
```diff
@@ -5,6 +5,9 @@ import { AsyncAPIDocument } from '@asyncapi/parser';
 
 declare global {
   interface Window {
+    // needed by monaco YAML plugin
+    monaco: typeof monacoAPI;
```
Comment (fmvilas) [inline src/react-app-env.d.ts:9]: `monaco` and `Monaco`? Can't we make it just one? Maybe make Studio get `monaco` instead?

File: src/services/monaco.service.ts [hunk 5]
```diff
@@ -89,8 +89,18 @@ export class MonacoService {
...
+    let monaco: typeof monacoAPI;
+
+    // JEST cannot bundle monaco-editor in test environment so we need to fetch that package from cdn
+    // in dev or production environment we will use bundled monaco-editor
+    if (process.env.NODE_ENV === 'test') {
+      monaco = await loader.init();
+    } else {
+      monaco = await import('monaco-editor');
+      loader.config({ monaco });
+    }
```
Comment (fmvilas) [inline src/services/monaco.service.ts:101]: I don't have much idea about the Monaco internals but every time I see code that needs to be changed to make tests work, it gives me shivers 😄 Isn't there really a way to make tests work without having to make this code aware of the tests?

Comment (magicmatatjahu) [reply src/services/monaco.service.ts:101]: I tried to map the imports to the appropriate modules, but this didn't work at all. Jest is building the project before testing (I was surprised) making monaco-editor mapping difficult. You can look for issues on github or even on stackoverflow where a lot of people have problems with this package on the jest side. I know that such ifs should not be, but this is the simplest solution to the problem.

File: package.json (hunk 3 above)
Comment (magicmatatjahu) [reply package.json:43]: Higher versions of `monaco-editor` have breaking changes, which are not compatible with given version of `monaco-yaml`. The latest version of `monaco-yaml` supports newer versions of monaco, but here again there are problems related to the worker on the yaml side. Each language in monaco is divided into 2 parts: the code on the main thread side, and the worker, which executes code not on the main thread, but on a separate thread to optimize the process (linting etc). Newer versions of `monaco-yaml` creates this worker differently, and to use it I would have to eject react-scripts and change the webpack. I want to fix this problem but not in this PR. Currently on the hosted studio `monaco-yaml` is not working fully (worker doesn't work) and I would prefer not to wait with these supports and just publish the fixed studio and then deal with dependencies.

Comment (fmvilas) [reply package.json:43]: That makes sense. Thanks or the explanation 👍

**Review by fmvilas — APPROVED** [review]
LGTM 👍 Thanks for the explanations!

**Pull request merged**